File: apps/ios/Flashcards/Flashcards/Review/Reaction/ReviewReactionState.swift
import Foundation

let reviewReactionMaximumActiveEvents: Int = 3

enum ReviewReactionMotionMode: Hashable, Sendable {
    case standard
    case reduced
}

enum ReviewReactionRating: CaseIterable, Hashable, Sendable {
    case again
    case hard
    case good
    case easy

    var debugIdentifier: String {
        String(describing: self)
    }
}

enum ReviewReactionVariant: CaseIterable, Hashable, Sendable {
    case againRainCloud
    case againTornado
    case againWindFace
    case againSnowflake
    case againSnailCrawl
    case againTurtle
    case againWiltedFlower
    case againSpider
    case againRat
    case againWormWiggle
    case hardTiger
    case hardTRex
    case hardShark
    case hardOxCharge
    case hardRacehorseGallop
    case hardSnake
    case hardVolcanoEruption
    case hardScorpion
    case hardPawPrints
    case hardRooster
    case goodOtter
    case goodOwl
    case goodRabbit
    case goodSeal
    case goodServiceDog
    case goodPoodle
    case goodChimpanzee
    case goodWhale
    case goodPeacock
    case goodPig
    case easySunrise
    case easySunriseOverMountains
    case easyRoseBloom
    case easyPeace
    case easyPlant
    case easyRainbowStreak
    case easyPhoenixRise
    case easyUnicornFlyby
    case fallbackCrownBounce

    var debugIdentifier: String {
        String(describing: self)
    }

    var animationDurationSeconds: Double {
        switch self {
        case .againWindFace:
            return 1.600
        case .hardTRex:
            return 1.550
        case .hardRacehorseGallop:
            return 1.20
        case .easySunriseOverMountains:
            return 1.200
        case .goodRabbit:
            return 1.333
        case .easyRoseBloom:
            return 2.40
        case .againSpider:
            return 2.400
        case .goodPeacock:
            return 1.333
        case .againTornado:
            return 1.45
        case .hardOxCharge:
            return 1.55
        case .hardScorpion:
            return 1.800
        case .hardPawPrints, .fallbackCrownBounce:
            return 1.65
        case .easyRainbowStreak:
            return 2.00
        case .hardVolcanoEruption:
            return 2.05
        case .againWiltedFlower:
            return 2.40
        case .goodSeal:
            return 2.567
        case .goodWhale:
            return 2.633
        case .againRat:
            return 2.633
        case .againSnailCrawl:
            return 2.70
        case .hardRooster:
            return 2.850
        case .goodPoodle:
            return 2.800
        case .goodOwl:
            return 2.833
        case .goodOtter, .goodServiceDog:
            return 3.000
        case .easyPeace:
            return 3.167
        case .hardShark:
            return 3.200
        case .againRainCloud, .hardSnake:
            return 3.267
        case .againTurtle:
            return 3.400
        case .goodPig:
            return 3.567
        case .goodChimpanzee:
            return 3.833
        case .easyPhoenixRise:
            return 3.933
        case .easyUnicornFlyby:
            return 3.80
        case .againSnowflake, .hardTiger, .easySunrise, .easyPlant:
            return 4.200
        case .againWormWiggle:
            return 4.267
        }
    }

    var cleanupDelayNanoseconds: UInt64 {
        let cleanupDelaySeconds = self.animationDurationSeconds + 0.08
        return UInt64(cleanupDelaySeconds * 1_000_000_000)
    }
}

struct ReviewReactionVariantDistributionEntry: Identifiable, Hashable, Sendable {
    let rating: ReviewReactionRating
    let variant: ReviewReactionVariant
    let weight: Int

    var id: String {
        "\(self.rating.debugIdentifier).\(self.variant.debugIdentifier)"
    }

    var probabilityPercent: Double {
        Double(self.weight) / Double(reviewReactionVariantTotalWeight(rating: self.rating)) * 100
    }
}

let allReviewReactionVariantDistributionEntries: [ReviewReactionVariantDistributionEntry] = [
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againRainCloud, weight: 32),
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againTornado, weight: 26),
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againWindFace, weight: 24),
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againSnowflake, weight: 18),
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againSnailCrawl, weight: 18),
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againTurtle, weight: 16),
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againWiltedFlower, weight: 12),
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againSpider, weight: 8),
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againRat, weight: 8),
    ReviewReactionVariantDistributionEntry(rating: .again, variant: .againWormWiggle, weight: 6),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardTiger, weight: 32),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardTRex, weight: 26),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardShark, weight: 22),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardOxCharge, weight: 20),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardRacehorseGallop, weight: 18),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardSnake, weight: 16),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardVolcanoEruption, weight: 14),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardScorpion, weight: 10),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardPawPrints, weight: 8),
    ReviewReactionVariantDistributionEntry(rating: .hard, variant: .hardRooster, weight: 8),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodOtter, weight: 32),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodOwl, weight: 28),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodRabbit, weight: 26),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodSeal, weight: 24),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodServiceDog, weight: 24),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodPoodle, weight: 20),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodChimpanzee, weight: 18),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodWhale, weight: 16),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodPeacock, weight: 12),
    ReviewReactionVariantDistributionEntry(rating: .good, variant: .goodPig, weight: 10),
    ReviewReactionVariantDistributionEntry(rating: .easy, variant: .easySunrise, weight: 34),
    ReviewReactionVariantDistributionEntry(rating: .easy, variant: .easySunriseOverMountains, weight: 34),
    ReviewReactionVariantDistributionEntry(rating: .easy, variant: .easyRoseBloom, weight: 30),
    ReviewReactionVariantDistributionEntry(rating: .easy, variant: .easyPeace, weight: 28),
    ReviewReactionVariantDistributionEntry(rating: .easy, variant: .easyPlant, weight: 26),
    ReviewReactionVariantDistributionEntry(rating: .easy, variant: .easyRainbowStreak, weight: 24),
    ReviewReactionVariantDistributionEntry(rating: .easy, variant: .easyPhoenixRise, weight: 18),
    ReviewReactionVariantDistributionEntry(rating: .easy, variant: .easyUnicornFlyby, weight: 12)
]

func reviewReactionVariantDistributionEntries(
    rating: ReviewReactionRating
) -> [ReviewReactionVariantDistributionEntry] {
    allReviewReactionVariantDistributionEntries.filter { entry in
        entry.rating == rating
    }
}

func reviewReactionVariantTotalWeight(
    rating: ReviewReactionRating
) -> Int {
    let entries = reviewReactionVariantDistributionEntries(rating: rating)
    precondition(!entries.isEmpty, "Review reaction distribution is missing rating \(rating.debugIdentifier).")

    var totalWeight = 0
    for entry in entries {
        precondition(entry.weight > 0, "Invalid review reaction weight for \(entry.id): \(entry.weight).")
        totalWeight += entry.weight
    }

    return totalWeight
}

func reviewReactionReadyVariantDistributionEntries(
    rating: ReviewReactionRating,
    readyVariants: Set<ReviewReactionVariant>
) -> [ReviewReactionVariantDistributionEntry] {
    reviewReactionVariantDistributionEntries(rating: rating).filter { entry in
        readyVariants.contains(entry.variant)
    }
}

func reviewReactionReadyVariantTotalWeight(
    rating: ReviewReactionRating,
    readyVariants: Set<ReviewReactionVariant>
) -> Int {
    let entries: [ReviewReactionVariantDistributionEntry] = reviewReactionReadyVariantDistributionEntries(
        rating: rating,
        readyVariants: readyVariants
    )

    var totalWeight: Int = 0
    for entry in entries {
        precondition(entry.weight > 0, "Invalid review reaction weight for \(entry.id): \(entry.weight).")
        totalWeight += entry.weight
    }

    return totalWeight
}

func reviewReactionAvailableVariantDistributionEntries(
    rating: ReviewReactionRating,
    availableVariants: Set<ReviewReactionVariant>
) -> [ReviewReactionVariantDistributionEntry] {
    reviewReactionVariantDistributionEntries(rating: rating).filter { entry in
        availableVariants.contains(entry.variant)
    }
}

func reviewReactionAvailableVariantTotalWeight(
    rating: ReviewReactionRating,
    availableVariants: Set<ReviewReactionVariant>
) -> Int {
    let entries: [ReviewReactionVariantDistributionEntry] = reviewReactionAvailableVariantDistributionEntries(
        rating: rating,
        availableVariants: availableVariants
    )

    var totalWeight: Int = 0
    for entry in entries {
        precondition(entry.weight > 0, "Invalid review reaction weight for \(entry.id): \(entry.weight).")
        totalWeight += entry.weight
    }

    return totalWeight
}

func reviewReactionCleanupDelayNanoseconds(
    variant: ReviewReactionVariant,
    motionMode: ReviewReactionMotionMode
) -> UInt64 {
    switch motionMode {
    case .standard:
        return variant.cleanupDelayNanoseconds
    case .reduced:
        return UInt64((ReviewReactionRenderer.reducedMotionDurationSeconds + 0.06) * 1_000_000_000)
    }
}

struct ReviewReactionEvent: Identifiable, Hashable, Sendable {
    let id: UUID
    let rating: ReviewReactionRating
    let variant: ReviewReactionVariant
}

func selectReviewReactionVariant(
    rating: ReviewReactionRating,
    roll: Int
) -> ReviewReactionVariant {
    let entries = reviewReactionVariantDistributionEntries(rating: rating)
    let totalWeight = reviewReactionVariantTotalWeight(rating: rating)
    precondition((0..<totalWeight).contains(roll), "Review reaction roll must be in 0..<\(totalWeight), received \(roll).")

    var cumulativeWeight = 0
    for entry in entries {
        cumulativeWeight += entry.weight
        if roll < cumulativeWeight {
            return entry.variant
        }
    }

    preconditionFailure("Review reaction distribution is missing rating \(rating.debugIdentifier) roll \(roll).")
}

func selectReadyReviewReactionVariant(
    rating: ReviewReactionRating,
    readyVariants: Set<ReviewReactionVariant>,
    roll: Int
) -> ReviewReactionVariant? {
    let entries: [ReviewReactionVariantDistributionEntry] = reviewReactionReadyVariantDistributionEntries(
        rating: rating,
        readyVariants: readyVariants
    )
    guard entries.isEmpty == false else {
        return nil
    }

    let totalWeight: Int = reviewReactionReadyVariantTotalWeight(
        rating: rating,
        readyVariants: readyVariants
    )
    precondition((0..<totalWeight).contains(roll), "Ready review reaction roll must be in 0..<\(totalWeight), received \(roll).")

    var cumulativeWeight: Int = 0
    for entry in entries {
        cumulativeWeight += entry.weight
        if roll < cumulativeWeight {
            return entry.variant
        }
    }

    preconditionFailure("Ready review reaction distribution is missing rating \(rating.debugIdentifier) roll \(roll).")
}

func selectAvailableReviewReactionVariant(
    rating: ReviewReactionRating,
    availableVariants: Set<ReviewReactionVariant>,
    roll: Int
) -> ReviewReactionVariant? {
    let entries: [ReviewReactionVariantDistributionEntry] = reviewReactionAvailableVariantDistributionEntries(
        rating: rating,
        availableVariants: availableVariants
    )
    guard entries.isEmpty == false else {
        return nil
    }

    let totalWeight: Int = reviewReactionAvailableVariantTotalWeight(
        rating: rating,
        availableVariants: availableVariants
    )
    precondition((0..<totalWeight).contains(roll), "Available review reaction roll must be in 0..<\(totalWeight), received \(roll).")

    var cumulativeWeight: Int = 0
    for entry in entries {
        cumulativeWeight += entry.weight
        if roll < cumulativeWeight {
            return entry.variant
        }
    }

    preconditionFailure("Available review reaction distribution is missing rating \(rating.debugIdentifier) roll \(roll).")
}

func makeReviewReactionRating(rating: ReviewRating) -> ReviewReactionRating {
    switch rating {
    case .again:
        return .again
    case .hard:
        return .hard
    case .good:
        return .good
    case .easy:
        return .easy
    }
}

func appendReviewReactionEvent(
    events: [ReviewReactionEvent],
    event: ReviewReactionEvent,
    maximumActiveEvents: Int
) -> [ReviewReactionEvent] {
    precondition(maximumActiveEvents > 0, "Review reactions require at least one active event slot.")

    let nextEvents = events + [event]
    guard nextEvents.count > maximumActiveEvents else {
        return nextEvents
    }

    return Array(nextEvents.suffix(maximumActiveEvents))
}

File: apps/android/feature/review/src/main/java/com/flashcardsopensourceapp/feature/review/reaction/ReviewReactionState.kt
package com.flashcardsopensourceapp.feature.review.reaction

import android.animation.ValueAnimator
import com.flashcardsopensourceapp.data.local.model.ReviewRating
import java.util.UUID
import kotlin.random.Random

internal const val reviewReactionMaximumActiveEvents: Int = 3

private const val reviewReactionReducedMotionDurationMillis: Int = 340

internal enum class ReviewReactionMotionMode {
    STANDARD,
    REDUCED
}

internal enum class ReviewReactionVariant {
    AGAIN_RAIN_CLOUD,
    AGAIN_TORNADO,
    AGAIN_WIND_FACE,
    AGAIN_SNOWFLAKE,
    AGAIN_SNAIL_CRAWL,
    AGAIN_TURTLE,
    AGAIN_WILTED_FLOWER,
    AGAIN_SPIDER,
    AGAIN_RAT,
    AGAIN_WORM_WIGGLE,
    HARD_TIGER,
    HARD_T_REX,
    HARD_SHARK,
    HARD_OX_CHARGE,
    HARD_RACEHORSE_GALLOP,
    HARD_SNAKE,
    HARD_VOLCANO_ERUPTION,
    HARD_SCORPION,
    HARD_PAW_PRINTS,
    HARD_ROOSTER,
    GOOD_OTTER,
    GOOD_OWL,
    GOOD_RABBIT,
    GOOD_SEAL,
    GOOD_SERVICE_DOG,
    GOOD_POODLE,
    GOOD_CHIMPANZEE,
    GOOD_WHALE,
    GOOD_PEACOCK,
    GOOD_PIG,
    EASY_SUNRISE,
    EASY_SUNRISE_OVER_MOUNTAINS,
    EASY_ROSE_BLOOM,
    EASY_PEACE,
    EASY_PLANT,
    EASY_RAINBOW_STREAK,
    EASY_PHOENIX_RISE,
    EASY_UNICORN_FLYBY,
    FALLBACK_CROWN_BOUNCE
}

internal val ReviewRating.reviewReactionDebugIdentifier: String
    get() = when (this) {
        ReviewRating.AGAIN -> "again"
        ReviewRating.HARD -> "hard"
        ReviewRating.GOOD -> "good"
        ReviewRating.EASY -> "easy"
    }

internal val ReviewReactionVariant.debugIdentifier: String
    get() = when (this) {
        ReviewReactionVariant.AGAIN_RAIN_CLOUD -> "againRainCloud"
        ReviewReactionVariant.AGAIN_TORNADO -> "againTornado"
        ReviewReactionVariant.AGAIN_WIND_FACE -> "againWindFace"
        ReviewReactionVariant.AGAIN_SNOWFLAKE -> "againSnowflake"
        ReviewReactionVariant.AGAIN_SNAIL_CRAWL -> "againSnailCrawl"
        ReviewReactionVariant.AGAIN_TURTLE -> "againTurtle"
        ReviewReactionVariant.AGAIN_WILTED_FLOWER -> "againWiltedFlower"
        ReviewReactionVariant.AGAIN_SPIDER -> "againSpider"
        ReviewReactionVariant.AGAIN_RAT -> "againRat"
        ReviewReactionVariant.AGAIN_WORM_WIGGLE -> "againWormWiggle"
        ReviewReactionVariant.HARD_TIGER -> "hardTiger"
        ReviewReactionVariant.HARD_T_REX -> "hardTRex"
        ReviewReactionVariant.HARD_SHARK -> "hardShark"
        ReviewReactionVariant.HARD_OX_CHARGE -> "hardOxCharge"
        ReviewReactionVariant.HARD_RACEHORSE_GALLOP -> "hardRacehorseGallop"
        ReviewReactionVariant.HARD_SNAKE -> "hardSnake"
        ReviewReactionVariant.HARD_VOLCANO_ERUPTION -> "hardVolcanoEruption"
        ReviewReactionVariant.HARD_SCORPION -> "hardScorpion"
        ReviewReactionVariant.HARD_PAW_PRINTS -> "hardPawPrints"
        ReviewReactionVariant.HARD_ROOSTER -> "hardRooster"
        ReviewReactionVariant.GOOD_OTTER -> "goodOtter"
        ReviewReactionVariant.GOOD_OWL -> "goodOwl"
        ReviewReactionVariant.GOOD_RABBIT -> "goodRabbit"
        ReviewReactionVariant.GOOD_SEAL -> "goodSeal"
        ReviewReactionVariant.GOOD_SERVICE_DOG -> "goodServiceDog"
        ReviewReactionVariant.GOOD_POODLE -> "goodPoodle"
        ReviewReactionVariant.GOOD_CHIMPANZEE -> "goodChimpanzee"
        ReviewReactionVariant.GOOD_WHALE -> "goodWhale"
        ReviewReactionVariant.GOOD_PEACOCK -> "goodPeacock"
        ReviewReactionVariant.GOOD_PIG -> "goodPig"
        ReviewReactionVariant.EASY_SUNRISE -> "easySunrise"
        ReviewReactionVariant.EASY_SUNRISE_OVER_MOUNTAINS -> "easySunriseOverMountains"
        ReviewReactionVariant.EASY_ROSE_BLOOM -> "easyRoseBloom"
        ReviewReactionVariant.EASY_PEACE -> "easyPeace"
        ReviewReactionVariant.EASY_PLANT -> "easyPlant"
        ReviewReactionVariant.EASY_RAINBOW_STREAK -> "easyRainbowStreak"
        ReviewReactionVariant.EASY_PHOENIX_RISE -> "easyPhoenixRise"
        ReviewReactionVariant.EASY_UNICORN_FLYBY -> "easyUnicornFlyby"
        ReviewReactionVariant.FALLBACK_CROWN_BOUNCE -> "fallbackCrownBounce"
    }

internal data class ReviewReactionVariantDistributionEntry(
    val rating: ReviewRating,
    val variant: ReviewReactionVariant,
    val weight: Int
) {
    val id: String
        get() = "${rating.reviewReactionDebugIdentifier}.${variant.debugIdentifier}"

    val probabilityPercent: Double
        get() = weight.toDouble() / reviewReactionVariantTotalWeight(rating = rating).toDouble() * 100.0
}

internal val allReviewReactionVariantDistributionEntries: List<ReviewReactionVariantDistributionEntry> = listOf(
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_RAIN_CLOUD, weight = 32),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_TORNADO, weight = 26),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_WIND_FACE, weight = 24),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_SNOWFLAKE, weight = 18),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_SNAIL_CRAWL, weight = 18),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_TURTLE, weight = 16),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_WILTED_FLOWER, weight = 12),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_SPIDER, weight = 8),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_RAT, weight = 8),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.AGAIN, variant = ReviewReactionVariant.AGAIN_WORM_WIGGLE, weight = 6),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_TIGER, weight = 32),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_T_REX, weight = 26),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_SHARK, weight = 22),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_OX_CHARGE, weight = 20),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_RACEHORSE_GALLOP, weight = 18),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_SNAKE, weight = 16),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_VOLCANO_ERUPTION, weight = 14),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_SCORPION, weight = 10),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_PAW_PRINTS, weight = 8),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.HARD, variant = ReviewReactionVariant.HARD_ROOSTER, weight = 8),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_OTTER, weight = 32),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_OWL, weight = 28),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_RABBIT, weight = 26),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_SEAL, weight = 24),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_SERVICE_DOG, weight = 24),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_POODLE, weight = 20),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_CHIMPANZEE, weight = 18),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_WHALE, weight = 16),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_PEACOCK, weight = 12),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.GOOD, variant = ReviewReactionVariant.GOOD_PIG, weight = 10),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.EASY, variant = ReviewReactionVariant.EASY_SUNRISE, weight = 34),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.EASY, variant = ReviewReactionVariant.EASY_SUNRISE_OVER_MOUNTAINS, weight = 34),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.EASY, variant = ReviewReactionVariant.EASY_ROSE_BLOOM, weight = 30),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.EASY, variant = ReviewReactionVariant.EASY_PEACE, weight = 28),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.EASY, variant = ReviewReactionVariant.EASY_PLANT, weight = 26),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.EASY, variant = ReviewReactionVariant.EASY_RAINBOW_STREAK, weight = 24),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.EASY, variant = ReviewReactionVariant.EASY_PHOENIX_RISE, weight = 18),
    ReviewReactionVariantDistributionEntry(rating = ReviewRating.EASY, variant = ReviewReactionVariant.EASY_UNICORN_FLYBY, weight = 12)
)

internal fun reviewReactionVariantDistributionEntries(
    rating: ReviewRating
): List<ReviewReactionVariantDistributionEntry> {
    return allReviewReactionVariantDistributionEntries.filter { entry: ReviewReactionVariantDistributionEntry ->
        entry.rating == rating
    }
}

internal fun reviewReactionVariantTotalWeight(rating: ReviewRating): Int {
    val entries: List<ReviewReactionVariantDistributionEntry> = reviewReactionVariantDistributionEntries(rating = rating)
    require(entries.isNotEmpty()) {
        "Review reaction distribution is missing a rating. rating=${rating.reviewReactionDebugIdentifier}"
    }

    var totalWeight = 0
    entries.forEach { entry: ReviewReactionVariantDistributionEntry ->
        require(entry.weight > 0) {
            "Review reaction weight must be positive. entryId=${entry.id} weight=${entry.weight}"
        }
        totalWeight += entry.weight
    }

    return totalWeight
}

internal data class ReviewReactionEvent(
    val id: String,
    val rating: ReviewRating,
    val variant: ReviewReactionVariant
)

internal fun selectReviewReactionVariant(
    rating: ReviewRating,
    roll: Int
): ReviewReactionVariant {
    val entries: List<ReviewReactionVariantDistributionEntry> = reviewReactionVariantDistributionEntries(rating = rating)
    val totalWeight: Int = reviewReactionVariantTotalWeight(rating = rating)
    require(roll in 0 until totalWeight) {
        "Review reaction roll must be in 0 until $totalWeight, received $roll."
    }

    var cumulativeWeight = 0
    for (entry in entries) {
        cumulativeWeight += entry.weight
        if (roll < cumulativeWeight) {
            return entry.variant
        }
    }

    error(
        "Review reaction distribution is missing a variant. " +
            "rating=${rating.reviewReactionDebugIdentifier} roll=$roll"
    )
}

internal fun appendReviewReactionEvent(
    events: List<ReviewReactionEvent>,
    event: ReviewReactionEvent,
    maximumActiveEvents: Int
): List<ReviewReactionEvent> {
    require(maximumActiveEvents > 0) {
        "Review reactions require at least one active event slot."
    }

    val nextEvents: List<ReviewReactionEvent> = events + event
    if (nextEvents.size <= maximumActiveEvents) {
        return nextEvents
    }

    return nextEvents.takeLast(n = maximumActiveEvents)
}

internal fun reviewReactionMotionModeFromAnimatorSettings(): ReviewReactionMotionMode {
    return if (ValueAnimator.areAnimatorsEnabled()) {
        ReviewReactionMotionMode.STANDARD
    } else {
        ReviewReactionMotionMode.REDUCED
    }
}

internal fun makeRandomReviewReactionEvent(rating: ReviewRating): ReviewReactionEvent {
    val totalWeight: Int = reviewReactionVariantTotalWeight(rating = rating)
    val roll: Int = Random.nextInt(from = 0, until = totalWeight)
    return ReviewReactionEvent(
        id = UUID.randomUUID().toString(),
        rating = rating,
        variant = selectReviewReactionVariant(rating = rating, roll = roll)
    )
}

internal fun reviewReactionAnimationDurationMillis(
    variant: ReviewReactionVariant,
    motionMode: ReviewReactionMotionMode
): Int {
    if (motionMode == ReviewReactionMotionMode.REDUCED) {
        return reviewReactionReducedMotionDurationMillis
    }

    return when (variant) {
        ReviewReactionVariant.AGAIN_WIND_FACE -> 1_600
        ReviewReactionVariant.HARD_T_REX -> 1_550
        ReviewReactionVariant.HARD_RACEHORSE_GALLOP -> 1_200
        ReviewReactionVariant.EASY_SUNRISE_OVER_MOUNTAINS -> 1_200
        ReviewReactionVariant.GOOD_RABBIT -> 1_333
        ReviewReactionVariant.EASY_ROSE_BLOOM -> 2_400
        ReviewReactionVariant.AGAIN_SPIDER -> 2_400
        ReviewReactionVariant.GOOD_PEACOCK -> 1_333
        ReviewReactionVariant.AGAIN_TORNADO -> 1_450
        ReviewReactionVariant.HARD_OX_CHARGE -> 1_550
        ReviewReactionVariant.HARD_SCORPION -> 1_800
        ReviewReactionVariant.HARD_PAW_PRINTS,
        ReviewReactionVariant.FALLBACK_CROWN_BOUNCE -> 1_650
        ReviewReactionVariant.EASY_RAINBOW_STREAK -> 2_000
        ReviewReactionVariant.HARD_VOLCANO_ERUPTION -> 2_050
        ReviewReactionVariant.AGAIN_WILTED_FLOWER -> 2_400
        ReviewReactionVariant.GOOD_SEAL -> 2_567
        ReviewReactionVariant.GOOD_WHALE -> 2_633
        ReviewReactionVariant.AGAIN_RAT -> 2_633
        ReviewReactionVariant.AGAIN_SNAIL_CRAWL -> 2_700
        ReviewReactionVariant.HARD_ROOSTER -> 2_850
        ReviewReactionVariant.GOOD_POODLE -> 2_800
        ReviewReactionVariant.GOOD_OWL -> 2_833
        ReviewReactionVariant.GOOD_OTTER,
        ReviewReactionVariant.GOOD_SERVICE_DOG -> 3_000
        ReviewReactionVariant.EASY_PEACE -> 3_167
        ReviewReactionVariant.HARD_SHARK -> 3_200
        ReviewReactionVariant.AGAIN_RAIN_CLOUD,
        ReviewReactionVariant.HARD_SNAKE -> 3_267
        ReviewReactionVariant.AGAIN_TURTLE -> 3_400
        ReviewReactionVariant.GOOD_PIG -> 3_567
        ReviewReactionVariant.GOOD_CHIMPANZEE -> 3_833
        ReviewReactionVariant.EASY_PHOENIX_RISE -> 3_933
        ReviewReactionVariant.EASY_UNICORN_FLYBY -> 3_800
        ReviewReactionVariant.AGAIN_SNOWFLAKE,
        ReviewReactionVariant.HARD_TIGER,
        ReviewReactionVariant.EASY_SUNRISE,
        ReviewReactionVariant.EASY_PLANT -> 4_200
        ReviewReactionVariant.AGAIN_WORM_WIGGLE -> 4_267
    }
}
